3DAmsterdam/Assets/Amsterdam3D/Scripts/Traffic Animation/Car.cs

```diff
@@ -0,0 +1,203 @@
+    private int currentRoadIndex = 0;
+    public int speed = 20;
+
+    public GameObject[] cars;
+
+    public bool debugCar = false;
+    private int findRoadLoopFrames = 0;
+
+    private int updateCarFrames = 0;
+    private float vehicleFrameSpeedCompensator = 1f;
+
+    private float carResetTimeStamp = 0.0f;
+    private float maxWaitingTime = 5f;
+
+    private float lastRecordedHeight = 0.0f;
+    // Start is called before the first frame update
+    void Start()
+    {
+        transform.position = currentRoad.roadPoints[0].pointCoordinates;
+        currentRoadIndex++;
+        // disables all car objects
+        foreach (GameObject car in cars)
+        {
+            car.SetActive(false);
+        }
+        // Chooses a random car out of all the car objects.
+        cars[Random.Range(0, cars.Length - 1)].SetActive(true);
+        speed = TrafficSimulator.Instance.carSpeed;
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        if (TrafficSimulator.Instance.enableCarSimulation && speed != TrafficSimulator.Instance.carSpeed)
+        {
+            speed = TrafficSimulator.Instance.carSpeed;
+        }
+        updateCarFrames++;
+        float DistanceToCar = Vector3.Distance(GenerateRoads.Instance.mainCameraTransform.position, transform.position);
+        if (DistanceToCar < TrafficSimulator.Instance.minimumCarRenderDistance)
+            vehicleFrameSpeedCompensator = 1f;
+        else if (DistanceToCar < TrafficSimulator.Instance.mediumCarRenderDistance && DistanceToCar > TrafficSimulator.Instance.minimumCarRenderDistance && updateCarFrames % 5 == 0)
+            vehicleFrameSpeedCompensator = 5f;
+        else if (DistanceToCar > TrafficSimulator.Instance.mediumCarRenderDistance && updateCarFrames % 10 == 0)
+            vehicleFrameSpeedCompensator = 10f;
+
+
+        if (DistanceToCar < TrafficSimulator.Instance.minimumCarRenderDistance ||
+             DistanceToCar < TrafficSimulator.Instance.mediumCarRenderDistance && DistanceToCar > TrafficSimulator.Instance.minimumCarRenderDistance && updateCarFrames % 5 == 0 ||
+            DistanceToCar > TrafficSimulator.Instance.mediumCarRenderDistance && updateCarFrames % 10 == 0)
+        {
+            if (!gameObject.activeSelf && TrafficSimulator.Instance.enableBoundsSimulation == false)
+            {
+                gameObject.SetActive(true);
+            }
+            if (TrafficSimulator.Instance.enableCarSimulation)
+            {
+                if (currentRoad != null)
+                {
+                    if (currentRoadIndex == 0)
+                    {
+                        transform.position = currentRoad.roadPoints[0].pointCoordinates;
+                        currentRoadIndex++;
+                    }
+
+                    RaycastHit hit;
+                    Vector3 temp = transform.position;
+                    temp.y = 50f;
+                    if (Physics.Raycast(temp, -Vector3.up, out hit, Mathf.Infinity))
```

Comment:
misschien dat het verstandiger is om hier Physics.RaycastNonAlloc te gebruiken, dan maak je zelf een buffer aan die je telkens kan herbruiken. anders doen alle autotjes best veel garbadge voor de garbadge collector aanmaken.
we zijn op zoek naar 1 hit ( ondergrond ) dus die ruimte kan al van tevoren gereserveerd worden